source/droid/gateway/compression.d

```diff
@@ -0,0 +1,37 @@
+module droid.gateway.compression;
+
+import droid.exception;
+import std.zlib,
+       std.stdio,
+       std.conv : to;
+
+enum CompressionType : string {
+    NONE        = "",
+    ZLIB        = "zlib",
+    ZLIB_STREAM = "zlib-stream"
+}
+class Decompressor {
+    string read(ubyte[] data) {
+        throw new DroidException("Compression type not supported!");
+    }
+}
+
+class ZLibStream : Decompressor {
+    const ulong[] ZLIB_SUFFIX = [0x0, 0x0, 0xFF, 0xFF];
+    UnCompress decompressor;
+
+    this() {
+        decompressor = new UnCompress(HeaderFormat.deflate);
+    }
+
+    override string read(ubyte[] data) {
+        if (data[$-4..$] != ZLIB_SUFFIX) {
+            throw new DroidException("ZLib-Stream compression enabled but invalid data was recieved!");
+        }
```

Comment:
This isn't an invalid case. I'm assuming the lack of length checking here is safe.

From the reference implementation:

```py
def on_websocket_message(msg):
  # always push the message data to your cache
  buffer.extend(msg)

  # check if the last four bytes are equal to ZLIB_SUFFIX
  if len(msg) < 4 or msg[-4:] != ZLIB_SUFFIX:
    return

  # if the message *does* end with ZLIB_SUFFIX,
  # get the full message by decompressing the buffers
  msg = inflator.decompress(buffer).decode('utf-8')
  buffer = bytearray()
```